---
title: "rse_jx_xplor"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(SummarizedExperiment)
```

```{r}
library(stringr) #for str_detect
library(readxl)
library(jaffelab)
library(biomaRt)
library(rtracklayer)
library(dplyr)

xls <- as.data.frame(read_excel('gene_list.xlsx'))

mart37 = useEnsembl(biomart="ensembl",GRCh=37, dataset = 'hsapiens_gene_ensembl')
#
attrs <- c('ensembl_gene_id', 'hgnc_symbol', 'external_gene_name', 'chromosome_name',  'start_position', 'end_position', 'strand', 'transcript_count') # 'ensembl_transcript_id', 'ccds')
gd37 <- getBM(mart=mart37, attributes=attrs, filters="ensembl_gene_id", values=xls$EnsemblID) 
# reorder to preserve the original xls order
gd37 <- gd37[match(xls$EnsemblID, gd37$ensembl_gene_id),]
## Gencode 25 corresponds to Ensembl 85
mart38 = useEnsembl(biomart="ensembl", dataset = 'hsapiens_gene_ensembl', version=85)
gd38 <- getBM(mart=mart38, attributes=attrs, filters="ensembl_gene_id", values=xls$EnsemblID) 
gmiss <- anti_join(gd37, gd38, by='ensembl_gene_id')
if (nrow(gmiss)>0) { # any genes recovered this way?
  gsymr <- getBM(mart=mart38, attributes=attrs, filters="hgnc_symbol", values=gmiss$hgnc_symbol) 
  gd38 <- rbind(gd38, gsymr)
}
```

```{r}
gmrg <- left_join(gd37, gd38, by=c('hgnc_symbol', 'chromosome_name'))
colnames(gmrg)=c('grch37_geneID', 'gsym', 'grch37_ext_gname', 'chr', 'grch37_start', 'grch37_end', 'grch37_strand', 'grch37_tcount', 'geneID', 'ext_gname', 'start', 'end', 'strand', 'tcount')
gmrg$grch37_strand=ifelse(gmrg$grch37_strand<0, '-', '+')
gmrg$strand=ifelse(gmrg$strand<0, '-', '+')
gmrg$chr <- paste0('chr', gmrg$chr)
## special case for gene 'ENSG00000261374', totally lost in GRch38

```



```{r}

rtl_gtf <- import("data/ann/gencode.v25.annotation.gtf")
ttbl <- as.data.frame(mcols(rtl_gtf)) %>% filter(type=='transcript') %>% 
  dplyr::select(gene_id, gene_type, gene_status, gene_name, transcript_id, transcript_type, transcript_name)


# get the number of transcripts for each gene:
gtbl <- ttbl %>% group_by(gene_id, gene_type, gene_status) %>% 
  summarise(num_transcripts=n(), transcripts=list(transcript_id))
# test: nrow(subset(gtbl, num_transcripts!=lengths(transcripts)))
# should be 0!

## ENSG00000261374 has no gene symbol, but in grch37.ensembl.org shows as 
## RP11-64K12.10 (Clone-based (Vega) gene), 
### https://grch37.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000261374
## has 1 non-coding single-exon transcript ENST00000567002.1 of length 1594
## maps to location  chr15:40374398-40375991 in GRCh38 coordinates

uix <- which(gmrg$grch37_geneID=='ENSG00000261374')
# check what genes are in that location on GRCh38/Gencode 25 ?
ovls <- findOverlaps(GRanges(seqnames="chr15", ranges=IRanges(start=40374398, end=40375991)) , rtl_gtf)
ghits <- rtl_gtf[subjectHits(ovls)]
# this shows that this genomic range is now fully contained in the 3' UTR (exon ENSE00000942471.5) 
# of transcript ENST00000267889.4 (DISP2-001) which is the only protein coding transcript of 
# the 4 transcripts of gene ENSG00000140323.5 (DISP2)
## -- list all transcripts for the overlapped gene(s) found:
rtl_gtf[which(mcols(rtl_gtf)$gene_id %in% ghits$gene_id & mcols(rtl_gtf)$type == 'transcript')]
### update the table with this info, somehow..
#|ENSE00000942471.5
gmrg[uix,c('geneID', 'ext_gname', 'start', 'end', 'strand', 'tcount')] <-
  list('ENSG00000140323', 'DISP2', 40374398, 40375991, '+', 4)
#note that 
write.csv(gmrg, 'genes_to_Gencode25.csv', row.names = F, quote = F)
```

```{r}
#jxann <- load("data/ann/junction_annotation_hg38_gencode_v25_main.rda")
# loads theJunctions GRanges object

#f2tx <- load("data/ann/feature_to_Tx_hg38_gencode_v25_main.rda")
##loads an allTx object
#exm <- load("data/ann/exonMaps_by_coord_hg38_gencode_v25.rda")
##loads coordToENSE, coordToTX, coordToEid GRanges object
```

Make genomic ranges out of our gene list
```{r}
ggr <- makeGRangesFromDataFrame(gmrg[,c('chr', 'strand', 'start', 'end', 'grch37_ext_gname', 'geneID', 'ext_gname')], keep.extra.columns = T)
colnames(mcols(ggr)) <- c('ogname', 'gene_id', 'gname')
```

load phenodata from the gene merged file, assuming that's the right data to use for all RSE objects!

```{r}
library(purrr)
library(recount)
#library(here)
load("count_data/RNAseq_Collection_postQC_n5536_11dataset_2021-01_geneRSE.Rdata", verbose = TRUE)
#pd <- as.data.frame(colData(rse_gene))
#phenodata <- pd
#table(pd$Dataset)
# --- skip next chunk 
load("phenodata.n2844.rda")
# ----- skip next chunk 

```


For the genes we might as well collect the data from the already loaded object:
SKIP THIS if you already loaded phenodata.n*.rda
```{r}
#rse_g_file <- "data/astellas_dg_hg38_rseGene_n263.rda"
#load(rse_g_file)
rse_gene <- rse_gene[ , rse_gene$Dataset %in% c('Astellas_DG','BrainSeq_Phase1','BrainSeq_Phase2_DLPFC',
                                                'BrainSeq_Phase2_HIPPO','BrainSeq_Phase3_Caudate','BrainSeq_Phase4and5')]
pd <- as.data.frame(colData(rse_gene))
#phenodata <- pd
table(pd$Dataset)
save(pd, file=paste0("phenodata.n",nrow(pd),".rda"))
```

SKIP THIS if you already loaded phenodata.n*.rda and the selected rse_gene was saved already
```{r}
rse_gene <- subsetByOverlaps(rse_gene, ggr)
xgrsub <- rowRanges(rse_gene)
xovls <- as.data.frame(findOverlaps(xgrsub, ggr))
if (nrow(xovls)!=length(xgrsub)) {
  message("WARNING: exons spanning multiple genes?!")
  stop();
  ## genes HSD3BP4 and HSD3BP5 are close will there be junctions overlapping both?
  #xovlg <- xovls %>% group_by(queryHits) %>% 
  #  summarize( genes=paste0(ggr[subjectHits]$ogname, collapse = '|'))
  #mcols(rowRanges(rse_exon_subset))$tgene <- xovlg$genes
}
mcols(rowRanges(rse_gene))$tgene <- ggr[xovls$subjectHits]$ogname
#rd <- as.data.frame(rowData(rse_tx))
# if ensemblID is not there, create it
#rowData(rse_gene)$ensemblID <- gsub('\\.\\d+$', '', rowData(rse_gene)$gencodeID)
#xmd <- mcols(rowRanges(rse_exon_subset))

## ? FIX this -- C22orf29 is also known as RTL10 (ENSG00000215012) and actually 
##                 overlaps exons of GNB1L (ENSG00000185838)!
#rse_jxn_subset <- rse_jxn_subset[ ! (row.names(rowData(rse_jxn_subset)) %in% 
#             rownames(to_remove)), ]

rse_gene <- rse_gene[rowData(rse_gene)$ensemblID %in% gmrg$geneID, ]
#rse_gene <- rse_gene[, rse_gene$Dx %in% c('Schizo', 'Control')]
table(rse_gene$Dx)
nrow(rse_gene)
save(rse_gene, file=paste0("rse_gene.25g.n",nrow(pd),".Rdata"))

```

```{r}
## load filenames
load(here("count_data","filenames.Rdata"), verbose = TRUE)
```


```{r}
## load exon data
rse_exon_l=list()
for (n in names(rse_filenames$exon)) {
  message("loading and filtering rse_exon for ",n)
  load(rse_filenames$exon[[n]]) # this loads a rse_exon
  if (n=="Astellas_DG") {
    colnames(rse_exon)=rse_exon$SAMPLE_ID
  } else {
    colnames(rse_exon)=sapply(rse_exon$SAMPLE_ID, paste, collapse=";")
  }
  
  rse_exon <- subsetByOverlaps(rse_exon, ggr)
  xgrsub <- rowRanges(rse_exon)
  xovls <- as.data.frame(findOverlaps(xgrsub, ggr))
  if (nrow(xovls)!=length(xgrsub)) {
      message("WARNING: exons spanning multiple genes?!")
      stop();
      ## genes HSD3BP4 and HSD3BP5 are close will there be junctions overlapping both?
  }
  mcols(rowRanges(rse_exon))$tgene <- ggr[xovls$subjectHits]$ogname
  rse_exon <- rse_exon[rowData(rse_exon)$ensemblID %in% gmrg$geneID, ]
  #rse_exon <- rse_exon[, rse_exon$Dx %in% c('Schizo', 'Control')]
  message(n, " : ", nrow(rse_exon))
  rse_exon_l[[n]] <- rse_exon
}
```
Now all the rse_exon objects are in rse_exon_l
```{r}

## Filter samples
original_ncol <- map_int(rse_exon_l, ncol)
exon_m <- map(rse_exon_l, ~colnames(.x) %in% rownames(pd))
## -- name it back rse_exon
rse_exon <- map2(rse_exon_l, exon_m, ~.x[,.y])
  
data.frame(original_ncol, new_ncol = map_int(rse_exon, ncol))  

#table(map_int(rse_exon_l, nrow))
#now merge rses in rse_exon_l and save
exon_rowRanges <- rowRanges(rse_exon[[1]])
rse_exon_counts <- map(rse_exon, ~assays(.x)$counts)

## combine count data
rse_exon_counts = do.call("cbind", rse_exon_counts)
dim(rse_exon_counts)
# [1] 571612   5536
rse_exon_counts <- rse_exon_counts[,match(rownames(pd), colnames(rse_exon_counts))]
all(colnames(rse_exon_counts) == rownames(pd))
## make rse_exon object
rse_exon = SummarizedExperiment(
  assays = list(counts = rse_exon_counts),
  rowRanges = exon_rowRanges, colData = pd)

rownames(rse_exon) <- rowData(rse_exon)$gencodeID
rowData(rse_exon)$meanExprs <- NULL
tempRpkm <- recount::getRPKM(rse_exon, "Length")
rowData(rse_exon)$meanExprs <- rowMeans(tempRpkm)
fn=paste0("rse_exon.25g.n",ncol(rse_exon_counts),".Rdata")
message("Writing file: ",fn)
save(rse_exon, file = fn)
```

Now do the same for transcripts
```{r}
## load and filter transcript data
rse_tx_l=list()
for (n in names(rse_filenames$tx)) {
  message("loading and filtering rse_tx for ",n)
  load(rse_filenames$tx[[n]]) # this loads a rse_tx
  if (n=="Astellas_DG") {
    colnames(rse_tx)=rse_tx$SAMPLE_ID
  } else {
    colnames(rse_tx)=sapply(rse_tx$SAMPLE_ID, paste, collapse=";")
  }
  
  rse_tx <- subsetByOverlaps(rse_tx, ggr)
  xgrsub <- rowRanges(rse_tx)
  xovls <- as.data.frame(findOverlaps(xgrsub, ggr))
  if (nrow(xovls)!=length(xgrsub)) {
     message("WARNING: transcripts spanning multiple genes?!")
     stop();
  }
  mcols(rowRanges(rse_tx))$tgene <- ggr[xovls$subjectHits]$ogname
  # ensemblID is not there, create it from gene_id (!)
  rowData(rse_tx)$ensemblID <- gsub('\\.\\d+$', '', rowData(rse_tx)$gene_id)
  rse_tx <- rse_tx[rowData(rse_tx)$ensemblID %in% gmrg$geneID, ]
  #table(rse_tx$Dx)
  message(n, " : ", nrow(rse_tx))
  message("    num samples:", dim(rse_tx)[2])
  rse_tx_l[[n]] <- rse_tx
}

```


```{r}
## Filter samples
original_ncol <- map_int(rse_tx_l, ncol)
tx_m <- map(rse_tx_l, ~colnames(.x) %in% rownames(pd))
rse_tx <- map2(rse_tx_l, tx_m, ~.x[,.y])

data.frame(original_ncol, new_ncol = map_int(rse_tx, ncol))

## Check # of tx
table(map_int(rse_tx, nrow))
# 198093 
# 11 
##Check that rownames are all equal
all(map_lgl(rse_tx, ~all(rownames(.x) == rownames(rse_tx[[1]]))))
# [1] TRUE
```

```{r}
## pick rowRanges data 
map_int(rse_tx, ~ncol(rowData(.x)))
```

```{r}
tx_rowRanges <- rowRanges(rse_tx[[1]])
## Extract count data
rse_tx_tpm <- map(rse_tx, ~assays(.x)$tpm)

## combine count data
rse_tx_tpm = do.call("cbind", rse_tx_tpm)
dim(rse_tx_tpm)
# [1] 198093   5536
table(rownames(pd) == colnames(rse_tx_tpm))

rse_tx_tpm <- rse_tx_tpm[,match(rownames(pd), colnames(rse_tx_tpm))]
all(colnames(rse_tx_tpm) == rownames(pd))

## make rse_exon object
rse_tx = SummarizedExperiment(
  assays = list(tpm = rse_tx_tpm),
  rowRanges = tx_rowRanges, colData = pd)

## save
fn=paste0("rse_tx.25g.n",ncol(rse_tx_tpm),".Rdata")
message("Writing file: ",fn)
save(rse_tx, file = fn)

```


Subsetting the junctions by overlapping each gene range on the same strand 
and not be assigned already another gene
```{r}
rse_jxn_l=list()
for (n in names(rse_filenames$jx)) {
  message("loading and filtering rse_jx for ",n)
  load(rse_filenames$jx[[n]]) # this loads a rse_tx
  if (n=="Astellas_DG") {
    colnames(rse_jxn)=rse_jxn$SAMPLE_ID
  } else {
    colnames(rse_jxn)=sapply(rse_jxn$SAMPLE_ID, paste, collapse=";")
  }
  ###TODO check - should we use ignore.strand=T ?
  rse_jxn <- subsetByOverlaps(rse_jxn, ggr)
  jgrsub <- rowRanges(rse_jxn)
  jovls <- as.data.frame(findOverlaps(ggr, jgrsub))

  ## genes HSD3BP4 and HSD3BP5 are too close and there will be junctions overlapping both
  jovlg <- jovls %>% group_by(subjectHits) %>% 
  summarize( genes=paste0(ggr[queryHits]$ogname, collapse = '|'))
  mcols(rowRanges(rse_jxn))$tgene <- jovlg$genes
  ## -> we should delete all junctions having !is.na(newGeneSymbol) & (newGeneSymbol!=tgene & tgene!=RP11-64K12.10
  ##   C22orf29 is also known as RTL10 (ENSG00000215012) and actually 
  ##                 overlaps exons of GNB1L (ENSG00000185838)!
  to_remove <- as.data.frame(rowData(rse_jxn)) %>% 
    filter(!is.na(newGeneSymbol) & !grepl('RP11-64K12.10', tgene) & 
             !(str_detect(tgene, fixed(newGeneSymbol)) | str_detect(newGeneSymbol, fixed(tgene) )) )

  rse_jxn <- rse_jxn[ ! (row.names(rowData(rse_jxn)) %in% rownames(to_remove)), ]
  #rse_jxn <- rse_jxn[, rse_jxn$Dx %in% c('Schizo', 'Control')]
  #table(rse_jxn$Dx)
  #table(rse_tx$Dx)
  message(n, " : ", nrow(rse_jxn))
  dim(rse_jxn)
  rse_jxn_l[[n]] <- rse_jxn
}
```

```{r}
#### Filter samples ####
original_ncol <- map_int(rse_jxn_l, ncol)
jxn_m <- map(rse_jxn_l, ~colnames(.x) %in% rownames(pd))
rse_jxn <- map2(rse_jxn_l, jxn_m, ~.x[,.y])
data.frame(original_ncol, new_ncol = map_int(rse_jxn, ncol))
```

```{r}
##TODO why do we have 2380 samples (and changing pd to it?!?q)
## Find all unique jxn
message("# Unique jxn")
jxn_all <- unique(unlist(map(rse_jxn, rownames)))
length(jxn_all)

message("Compile new rowData")
rowData_new <- unlist( as(map(rse_jxn, rowRanges),"GRangesList"))
names(rowData_new) <- ss(names(rowData_new),"\\.",2)
rowData_new <- unique(rowData_new)
length(rowData_new) == length(jxn_all)

## Add missing jxn to each set
jxn_counts <- map(rse_jxn, function(x){
  counts_temp <- assays(x)$counts
  # message(class(counts_temp))
  # mode(counts_temp) <- "integer"
  new_jxn <- jxn_all[!jxn_all %in% rownames(counts_temp)]
  message(length(new_jxn))
  counts_new <- matrix(data = 0, nrow = length(new_jxn), ncol = ncol(x), dimnames = list(new_jxn, colnames(x)))
  # counts_new <- Matrix(data = 0, nrow = length(new_jxn), ncol = ncol(x), dimnames = list(new_jxn, colnames(x)), sparse = TRUE)
  return(rbind(counts_temp, counts_new))
})
message("Size of matrix list")
print(object.size(jxn_counts),units = "auto")

```
```{r}
jxn_counts <- do.call("cbind", jxn_counts)
message("Size of combined matrix")
dim(jxn_counts)
print(object.size(jxn_counts),units = "auto")

pd <- pd[colnames(jxn_counts),]
dim(pd)

```
```{r}
all(colnames(jxn_counts) == rownames(pd))

rse_jxn = SummarizedExperiment(
  assays = list(counts = jxn_counts),
  rowRanges = rowData_new, colData = pd)
fn=paste0("rse_jxn.25g.n",ncol(jxn_counts),".Rdata")
message("Writing file: ",fn)
save(rse_jxn, file = fn)
```


```{r}
rse_jxn_file <- "data/astellas_dg_hg38_rseJxn_n263.rda"
load(rse_jxn_file)
## retrieve all junctions for each of the genomic ranges
# jovls <- findOverlaps(ggr, jgr, ignore.strand=T)
## subset rowData accordingly:
# jrd_hits <- jrd[subjectHits(jovls),]
## subset genomic ranges:
## jgr[subjectHits(jovls)]
###TODO check - should we use ignore.strand=T ?
rse_jxn <- subsetByOverlaps(rse_jxn, ggr)
jgrsub <- rowRanges(rse_jxn)
jovls <- as.data.frame(findOverlaps(ggr, jgrsub))

## genes HSD3BP4 and HSD3BP5 are too close and there will be junctions overlapping both
jovlg <- jovls %>% group_by(subjectHits) %>% 
  summarize( genes=paste0(ggr[queryHits]$ogname, collapse = '|'))
mcols(rowRanges(rse_jxn))$tgene <- jovlg$genes
## -> we should delete all junctions having !is.na(newGeneSymbol) & (newGeneSymbol!=tgene & tgene!=RP11-64K12.10
## TODO FIX this -- C22orf29 is also known as RTL10 (ENSG00000215012) and actually 
##                 overlaps exons of GNB1L (ENSG00000185838)!
to_remove <- as.data.frame(rowData(rse_jxn)) %>% 
    filter(!is.na(newGeneSymbol) & !grepl('RP11-64K12.10', tgene) & 
             !(str_detect(tgene, fixed(newGeneSymbol)) | str_detect(newGeneSymbol, fixed(tgene) )) )

rse_jxn <- rse_jxn[ ! (row.names(rowData(rse_jxn)) %in% rownames(to_remove)), ]
#rse_jxn <- rse_jxn[, rse_jxn$Dx %in% c('Schizo', 'Control')]
table(rse_jxn$Dx)
##TODO - save rse_jxn to get it ready for merging later

```

----------- END HERE -----------------

Finally, genes themselves
```{r}
stop() #done already above, don't include this in the loop

rse_g_file <- "data/astellas_dg_hg38_rseGene_n263.rda"
load(rse_g_file)
rse_gene <- subsetByOverlaps(rse_gene, ggr)
xgrsub <- rowRanges(rse_gene)
xovls <- as.data.frame(findOverlaps(xgrsub, ggr))
if (nrow(xovls)!=length(xgrsub)) {
  message("WARNING: exons spanning multiple genes?!")
  stop();
  ## genes HSD3BP4 and HSD3BP5 are close will there be junctions overlapping both?
  #xovlg <- xovls %>% group_by(queryHits) %>% 
  #  summarize( genes=paste0(ggr[subjectHits]$ogname, collapse = '|'))
  #mcols(rowRanges(rse_exon_subset))$tgene <- xovlg$genes
}
mcols(rowRanges(rse_gene))$tgene <- ggr[xovls$subjectHits]$ogname
#rd <- as.data.frame(rowData(rse_tx))
# if ensemblID is not there, create it
#rowData(rse_gene)$ensemblID <- gsub('\\.\\d+$', '', rowData(rse_gene)$gencodeID)
#xmd <- mcols(rowRanges(rse_exon_subset))

## ? FIX this -- C22orf29 is also known as RTL10 (ENSG00000215012) and actually 
##                 overlaps exons of GNB1L (ENSG00000185838)!
#rse_jxn_subset <- rse_jxn_subset[ ! (row.names(rowData(rse_jxn_subset)) %in% 
#             rownames(to_remove)), ]

rse_gene <- rse_gene[rowData(rse_gene)$ensemblID %in% gmrg$geneID, ]
rse_gene <- rse_gene[, rse_gene$Dx %in% c('Schizo', 'Control')]
table(rse_gene$Dx)
nrow(rse_gene)
```


